gcp/iap-ingress/overlays/application/params.yaml

```diff
@@ -7,3 +7,5 @@ varReference:
   kind: StatefulSet
 - path: spec/template/metadata/labels/app.kubernetes.io\/instance
   kind: StatefulSet
+- path: spec/template/metadata/labels/app.kubernetes.io\/instance
+  kind: Job
```

Comment:
I'm trying to understand this. Just correct me here if I'm not getting this right - the Application overlay should label all resources belonging to `iap-ingress`. So here all resources:
- BackendConfig.yaml
- CloudEndpoint
- ClusterRoleBinding.yaml
- ClusterRole
- ConfigMap
- Deployment
- Ingress
- Policy
- ServiceAccount
- Service

Should get labelled with `app.kubernetes.io/instance`.

And likewise, for all other application/params.yaml in this PR. This, of course, doesn't need to be a part of this PR.

Reply:
Ok, I think I figured this out.

Looks like the labels not in metadata/labels require a varReference.

Reply:
@krishnadurai @richardsliu @jlewi @yanniszark 
here are some details. I think your understanding is correct, but I'll explain as well:

Adding the application overlay creates an Application CR for that manifest. Applications will track their components and report their status. This is key for e2e tests since we now have a better approach to health monitoring than just checking deployments ([the current e2e tests](https://github.com/kubeflow/kubeflow/blob/master/testing/kfctl/kf_is_ready_test.py#L35)). 

We add labels to all resources per application - for example in [centraldashboard](https://github.com/kubeflow/manifests/blob/master/common/centraldashboard/overlays/application/kustomization.yaml#L20). Notice the $(generateName) in the labels. Each application has a unique name which kfctl will generate by adding a suffix to the name. For example centraldashboard will have a name like [centraldashboard-akdfjls](https://github.com/kubeflow/manifests/blob/master/kfdef/kfctl_gcp_iap.yaml#L93).

When kustomize sees commonLabels it adds these to all resources labels. This is [recommended](https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/#labels) by kubernetes. Kustomize will also add selectors to resources like deployments, statefulsets, daemonsets, etc. Unfortunately kustomize [isn't doing var replacement](https://github.com/kubernetes-sigs/kustomize/issues/1585) when it adds selectors. 

Additionally adding the instance label to all resources allows us to delete all resources including cluster wide resources even if we're not adding that resource to the application.componentKinds. Deletion of cluster wide resources has been broken for a while in kubeflow so this is a high priority.

Finally this is needed for app upgrades.

Reply:
Thanks @kkasravi!
I think we could capture the application CR part in docs too.